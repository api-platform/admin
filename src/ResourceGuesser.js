import React from 'react';
import { Resource } from 'react-admin';
import PropTypes from 'prop-types';
import ListGuesser from './ListGuesser';
import CreateGuesser from './CreateGuesser';
import EditGuesser from './EditGuesser';
import ShowGuesser from './ShowGuesser';

const ResourceGuesser = ({ list, edit, create, show, ...props }) => (
  <Resource
    {...props}
    create={create || CreateGuesser}
    edit={edit || EditGuesser}
    list={list || ListGuesser}
    show={show || ShowGuesser}
  />
);

ResourceGuesser.propTypes = {
  resource: PropTypes.string.isRequired,
};

export default ResourceGuesser;
