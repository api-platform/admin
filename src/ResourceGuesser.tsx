import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import {
  Resource,
  useResourceDefinition,
  useResourceDefinitionContext,
} from 'react-admin';
import type { ResourceDefinition, ResourceProps } from 'react-admin';
import ListGuesser from './ListGuesser.js';
import CreateGuesser from './CreateGuesser.js';
import EditGuesser from './EditGuesser.js';
import ShowGuesser from './ShowGuesser.js';
import Introspecter from './Introspecter.js';
import type {
  IntrospectedResourceGuesserProps,
  ResourceGuesserProps,
} from './types.js';

export const IntrospectedResourceGuesser = ({
  resource,
  schema,
  schemaAnalyzer,
  list = ListGuesser,
  edit = EditGuesser,
  create = CreateGuesser,
  show = ShowGuesser,
  ...props
}: IntrospectedResourceGuesserProps) => {
  const { register } = useResourceDefinitionContext();
  const registeredDefinition = useResourceDefinition({ resource });

  let hasList = false;
  let hasEdit = false;
  let hasCreate = false;
  let hasShow = false;
  schema.operations?.forEach((operation) => {
    if (operation.type === 'list') {
      hasList = true;
    }
    if (operation.type === 'edit') {
      hasEdit = true;
    }
    if (operation.type === 'create') {
      hasCreate = true;
    }
    if (operation.type === 'show') {
      hasShow = true;
    }
  });

  useEffect(() => {
    if (
      registeredDefinition.hasList !== hasList ||
      registeredDefinition.hasEdit !== hasEdit ||
      registeredDefinition.hasCreate !== hasCreate ||
      registeredDefinition.hasShow !== hasShow
    ) {
      register({
        name: resource,
        icon: props.icon,
        options: props.options,
        hasList,
        hasEdit,
        hasCreate,
        hasShow,
      });
    }
  }, [
    register,
    resource,
    props.icon,
    props.options,
    hasList,
    hasEdit,
    hasCreate,
    hasShow,
    registeredDefinition,
  ]);

  return (
    <Resource
      {...props}
      name={resource}
      create={create}
      edit={edit}
      list={list}
      show={show}
    />
  );
};

const ResourceGuesser = ({ name, ...props }: ResourceGuesserProps) => (
  <Introspecter
    component={IntrospectedResourceGuesser}
    resource={name}
    {...props}
  />
);

ResourceGuesser.raName = 'Resource';

ResourceGuesser.registerResource = (
  props: ResourceProps,
): ResourceDefinition => ({
  name: props.name,
  icon: props.icon,
  options: props.options,
  hasList: true,
  hasEdit: true,
  hasCreate: true,
  hasShow: true,
});

ResourceGuesser.propTypes = {
  name: PropTypes.string.isRequired,
};

export default ResourceGuesser;
