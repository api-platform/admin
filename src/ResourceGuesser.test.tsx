import React from 'react';
import ResourceGuesser from './ResourceGuesser';
import ShallowRenderer from 'react-test-renderer/shallow';

describe('<ResourceGuesser />', () => {
  const renderer = new ShallowRenderer();

  test('renders with create', () => {
    const CustomCreate = () => null;

    const tree = renderer.render(
      <ResourceGuesser name="dummy" create={CustomCreate} />,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders without create', () => {
    const tree = renderer.render(<ResourceGuesser name="dummy" />);

    expect(tree).toMatchSnapshot();
  });

  test('renders with edit', () => {
    const CustomEdit = () => null;

    const tree = renderer.render(
      <ResourceGuesser name="dummy" edit={CustomEdit} />,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders without edit', () => {
    const tree = renderer.render(<ResourceGuesser name="dummy" />);

    expect(tree).toMatchSnapshot();
  });

  test('renders with list', () => {
    const CustomList = () => null;

    const tree = renderer.render(
      <ResourceGuesser name="dummy" list={CustomList} />,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders without list', () => {
    const tree = renderer.render(<ResourceGuesser name="dummy" />);

    expect(tree).toMatchSnapshot();
  });

  test('renders with show', () => {
    const CustomShow = () => null;

    const tree = renderer.render(
      <ResourceGuesser name="dummy" show={CustomShow} />,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders without show', () => {
    const tree = renderer.render(<ResourceGuesser name="dummy" />);

    expect(tree).toMatchSnapshot();
  });
});
