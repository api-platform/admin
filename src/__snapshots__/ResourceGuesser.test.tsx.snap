// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResourceGuesser /> renders with create 1`] = `
<Introspecter
  component={[Function]}
  create={[Function]}
  resource="dummy"
/>
`;

exports[`<ResourceGuesser /> renders with edit 1`] = `
<Introspecter
  component={[Function]}
  edit={[Function]}
  resource="dummy"
/>
`;

exports[`<ResourceGuesser /> renders with list 1`] = `
<Introspecter
  component={[Function]}
  list={[Function]}
  resource="dummy"
/>
`;

exports[`<ResourceGuesser /> renders with show 1`] = `
<Introspecter
  component={[Function]}
  resource="dummy"
  show={[Function]}
/>
`;

exports[`<ResourceGuesser /> renders without create 1`] = `
<Introspecter
  component={[Function]}
  resource="dummy"
/>
`;

exports[`<ResourceGuesser /> renders without edit 1`] = `
<Introspecter
  component={[Function]}
  resource="dummy"
/>
`;

exports[`<ResourceGuesser /> renders without list 1`] = `
<Introspecter
  component={[Function]}
  resource="dummy"
/>
`;

exports[`<ResourceGuesser /> renders without show 1`] = `
<Introspecter
  component={[Function]}
  resource="dummy"
/>
`;
